src/frontend/flight/sync/utils.rs

```diff
@@ -0,0 +1,425 @@
+                chain_count += 1;
+            }
+        };
+    }
+
+    // Transpose the changed rows so that we obtain the indices for each column role
+    let rows = column_rows.len();
+    let (mut old_pks, mut new_pks, mut changed_rows) = (
+        Vec::with_capacity(rows),
+        Vec::with_capacity(rows),
+        VecDeque::with_capacity(rows),
+    );
+    for (old_pk, new_pk, changed_row) in column_rows {
+        old_pks.push(old_pk);
+        new_pks.push(new_pk);
+        for (idx, changed_row) in changed_row.iter().enumerate() {
+            match changed_rows.get_mut(idx) {
+                None => changed_rows.push_back(vec![*changed_row]),
+                Some(v) => v.push(*changed_row),
+            };
+        }
+    }
+
+    // Construct the actual index arrays for each column
+    let old_pk_indices = UInt64Array::from(old_pks);
+    let new_pk_indices = UInt64Array::from(new_pks);
+    let changed_indices = changed_rows
+        .iter()
+        .map(|rows| UInt64Array::from(rows.clone()))
+        .collect::<Vec<_>>();
+    let mut changed_pos = 0;
+    let mut indices: HashMap<usize, &UInt64Array> = HashMap::new();
+    sync_schema.columns().enumerate().for_each(|(col_id, col)| {
+        match col.role() {
+            ColumnRole::OldPk => {
+                indices.insert(col_id, &old_pk_indices);
+            }
+            ColumnRole::NewPk => {
+                indices.insert(col_id, &new_pk_indices);
+            }
+            ColumnRole::Changed => {
+                // Insert the indices for both this column...
+                indices.insert(col_id, &changed_indices[0]);
+
+                // ... as well as for the actual changed Value column
+                let changed_col_id = schema
+                    .index_of(
+                        sync_schema
+                            .column(col.name(), ColumnRole::Value)
+                            .unwrap()
+                            .field()
+                            .name(),
+                    )
+                    .expect("Field exists");
+                indices.insert(changed_col_id, &changed_indices[0]);
+                changed_pos += 1;
+            }
+            ColumnRole::Value => {
+                indices.entry(col_id).or_insert(&new_pk_indices);
+            }
+        }
+    });
+
+    // Finally take the designated rows from each old PK, new PK, value and changed columns from the
+    // batch
+    let batch = RecordBatch::try_new(
+        schema.clone(),
+        (0..schema.fields().len())
+            .map(|col_id| {
+                Ok(take(
+                    batch.columns()[col_id].as_ref(),
+                    indices.get(&col_id).expect("Indices present"),
+                    None,
+                )?)
+            })
+            .collect::<Result<_>>()?,
+    )?;
+
+    Ok(batch)
+}
+
+// Generates a pruning qualifier for the table based on the PK columns and the min/max values
+// in the data sync entries.
```

Comment:
I don't quite understand the point of this, like what is the benefit? The comment doesn't help me much

Reply:
The goal is to reduce the amount of data needed to be scanned/streamed/joined with from the base table.

Basically we only need to touch those partition which might have the extreme (min/max) PK values that are contained in the sync data.

Reply:
Got it, changing the doc.

> Aside: if my entry has two changes to PK=1 and PK=infinity, will this make me read the entire Delta table?

Yup, I don't think there's a way around it.

Reply:
I think we can still work around it though - by manually reading the Delta metadata and cross-referencing it against each PK we know we have changes for, but that sounds like more of a future optimization to me.

Reply:
But delta-rs already does that for us automatically, i.e. the scan uses the filter from this function to prune the partitions. So if we have PK=1..infinity (I'm assuming you mean here a really large number encompassing all partitions) then we must hit all partitions otherwise we're not guaranteeing idempotency (e.g. we may double-insert)?

Reply:
I was thinking just two changes, one hitting partition #1, one hitting the last partition, not a range of changes spanning all partitions. Surely in this case we can optimize by only rewriting these two partitions?

Reply:
Ah, yeah that should be possible. 

Not sure how complicated it would be though, naively in the general case we'd need to have a qualifier per change/sync row and consequently a multiple scans/joins. 

The less naive approach would be to segment all changes and group them by partitions they hit, and then only scan/join per group.